### PR TITLE
[Bugfix] Fix counts within Exam Checklist and Locking bugs with programming exercises

### PR DESCRIPTION
**Pull request by nicolasruscher**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
A few bugs were discovered recently and we aim to fix them in this PR.

1. Currently test runs are still included in the counts, which should not be the case. We removed them by using the correct query

2. A index out of bounds exception: When the user tried to access a result of the first correction round, when the result of the second correction round did already existed aswell this exception was thrown due to an erroneous check. `>=` was needed instead of `==`.

3. When the only result (a manual one) of a programmingSubmission was cancelled, the submission was deleted as well. At some point in the second correction development, a new submission was created for each new manual assessment. Since this is not the case anymore we mustn't delete the submission anymore.

### Steps for Testing

1. Make sure that within the exam checklist no counts (especially within the assessment row) include the test runs.
2. go to an exam, to a programming exercise -> exercise-groups -> submissions. In this dashboard click on "open assessment" for the first correction round, when the "open assessment" button is also visible in the second column. This should open the first result correctly.
3.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java [hunk 1]
```diff
@@ -111,15 +111,12 @@
      *         exercise due date at all
      */
     @Query("""
-            SELECT COUNT (DISTINCT p) FROM StudentParticipation p
+            SELECT COUNT (DISTINCT p) FROM StudentParticipation p join  p.submissions s
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java:114]: Code looks good, but I find it a little bit strange that you are counting the participations here and not the submission, but I guess it is fine if it works :) Gj

**Review by stefanwaldhauser — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java (hunk 1 above)
Comment (nicolasruscher) [reply src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java:114]: Its legacy code, I only rewrote this part of the query to include performance improvements 😄. We will refactor some of these old queries in a followup.

**Review by FuchsDominik — APPROVED** [review]
Screenshot 2021-02-19 at 11 30 28
Looks good I think. Two submitted exams and one test run done with each 2 exercises and 4 assessments are To-Do.

**Review by SimonEntholzer — APPROVED** [review]
Code looks good, testing locally worked. The multiple created testruns aren't counted. The code is good, too, and we will refactor some parts soon anyway :)

**Review by tiffca — DISMISSED** [review]
Counts look correct

**Review by airelawaleria — DISMISSED** [review]

**Review by simonzilker — APPROVED** [review]
No issues found :)

**Review by JohannesWeiss — DISMISSED** [review]
Works as intended on ts1

**Review by EneaGore — APPROVED** [review]
Seems fine on ts1

**Review by marvinchimeka — APPROVED** [review]
No issues found

**Review by krusche — APPROVED** [review]

**Pull request merged**